### ISSUE TITLE
Pyright does not correctly typecheck __setattr__ when calling through `=`

### ISSUE THREAD

**Issue by tylerhou** [microsoft/pyright#3656]
**Describe the bug**

When defining a class with `__setattr__`, Pyright correctly type checks direct calls to `__setattr__`, but does not type check indirect calls to `__setattr__` through setting an attribute on an instance.

**To Reproduce**

```python
# test.py
class A:
    def __setattr__(self, key, val: int):
        pass

A().a = '1'  # this line
```

Run `pyright test.py`.
**Expected behavior**

Pyright should flag the marked `# this line` with a type error.

**Additional context**

mypy flags the line as a type error. Pytype does not.

```
test ❯ mypy test.py
test.py:5: error: Incompatible types in assignment (expression has type "str", variable has type "int")
Found 1 error in 1 file (checked 1 source file)
```

**Comment by erictraut** [comment]
Thanks for the bug report. I'm able to repro the bug, and I understand the underlying cause. I'll work on a fix.

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by tylerhou** [comment]
(Will be fixed by commit 3ec06addd09bf397f318c1317e0bded8e0258f25.)

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.258, which I just published. It will also be included in the next release of pylance.
